packages/dm-core-plugins/src/file/FilePlugin.tsx

```diff
@@ -0,0 +1,202 @@
+import styled from 'styled-components'
+import { useRegistryContext } from '../form/RegistryContext'
+
+export const ErrorGroup = styled.div`
+  display: flex;
+  flex-direction: column;
+  border: 1px solid rgba(213, 18, 18, 0.71);
+  border-radius: 5px;
+  padding: 20px 20px;
+  background-color: #f6dfdf;
+`
+
+const UploadFileButton = (props: any) => {
+  const { onUpload } = props
+  const textInput = useRef<HTMLInputElement>(null)
+  const [error, setError] = useState<string | undefined>()
+  const [loading, setLoading] = useState<boolean>(false)
+  const dmssAPI = useDMSS()
+  const { idReference } = useRegistryContext()
+  const dataSourceId = idReference?.split('/', 2)[0]
+
+  useEffect(() => setError(undefined))
+
+  function handleUpload(event: any): void {
+    setError('')
+    const file = event.target.files[0]
+    setLoading(true)
+    const blobId = crypto.randomUUID()
+    dmssAPI
+      .blobUpload({
+        dataSourceId: dataSourceId,
+        file: file,
+        blobId: blobId,
+      })
+      .then(() => {
+        const reference = {
+          referenceType: 'link',
+          address: `$${blobId}`,
+          type: 'dmss://system/SIMOS/Reference',
+        }
+        onUpload(file, reference)
+      })
+      .catch((error: AxiosError<any>) => {
+        console.log(error)
+        const errorResponse =
+          typeof error.response?.data == 'object'
+            ? error.response?.data?.message
+            : error.response?.data
+        const errorMessage = errorResponse || 'Failed to upload file'
+        setError(errorMessage)
+      })
+      .finally(() => setLoading(false))
+  }
+
+  if (error)
+    return (
+      <ErrorGroup>
+        <b>Error</b>
+        <b>
+          Failed uploading file...
+          <div>
+            <code>{error}</code>
+          </div>
+        </b>
+      </ErrorGroup>
+    )
+
+  return (
+    <>
+      <input
+        type="file"
+        ref={textInput}
+        style={{ display: 'none' }}
+        onChange={(event: ChangeEvent<HTMLInputElement>) => handleUpload(event)}
+      />
+      {loading ? (
+        <Button style={{ margin: '0 10px' }}>
+          <Progress.Dots />
+        </Button>
+      ) : (
+        <Button
+          onClick={() => textInput?.current?.click()}
+          style={{ margin: '0 10px' }}
+        >
+          Upload
+        </Button>
+      )}
+      {error && <p style={{ color: 'red' }}>{error}</p>}
+    </>
+  )
+}
+
+const DownloadButton = (props: any) => {
+  const { entity, dataSourceId } = props
+  const [fileUrl, setFileUrl] = useState('')
+  const [loading, setLoading] = useState(true)
+  const [error, setError] = useState<string | null>(null)
+
+  const dmssAPI = useDMSS()
+
+  useEffect(() => {
+    setError(null)
+    const options: AxiosRequestConfig = {
+      responseType: 'blob',
+    }
+    dmssAPI
+      .blobGetById(
+        {
+          dataSourceId: dataSourceId,
+          blobId: entity.content['address'],
```

Comment:
`entity.content['address']` assumes that entity sent in as props is of a special type, but i'm not sure which. Is it a link reference?

Can rename entity prop to something like "LinkReferenceEntity" or whateter type its supposed to be

Reply:
good idea, renamed to fileEntity